README.md

# Easy train

## docker without gpu
docker run -it --rm -v C:\obara\pytorch-AdaIN:/project docker040_neuro /bin/bash
## nvidia docker
sudo docker run --gpus all -it --shm-size=4G -v /home/hideto.obara/pytorch-AdaIN/:/project neuro /bin/bash
## prepare decoder
nohup python3 train.py --content_dir input/content/ --style_dir input/style > train.out
## make abstracter,corrector
nohup python3 train.py --content_dir input/content/ --style_dir input/style --decoder models/decoder_default.pth.tar > train.out
## generate
python3 test.py --content input/content/lenna.jpg --decoder models/decoder_default.pth.tar --abstracter models/abstracter_default.pth.tar --corrector models/corrector_default.pth.tar
python3 test.py --content input/content/lenna.jpg --decoder models/decoder_saya.pth.tar --abstracter models/abstracter_saya.pth.tar --corrector models/corrector_saya.pth.tar

# pytorch-AdaIN

This is an unofficial pytorch implementation of a paper, Arbitrary Style Transfer in Real-time with Adaptive Instance Normalization [Huang+, ICCV2017].
I'm really grateful to the [original implementation](https://github.com/xunhuang1995/AdaIN-style) in Torch by the authors, which is very useful.

![Results](results.png)

## Requirements
- Python 3.5+
- PyTorch 0.4+
- TorchVision
- Pillow

(optional, for training)
- tqdm
#- TensorboardX

## Usage

### Download models
This command will download a pre-trained decoder as well as a modified VGG-19 network.
```
bash models/download_models.sh
```

### Convert models
This command will convert the models for Torch to the models for PyTorch.
```
python torch_to_pytorch.py --model models/vgg_normalised.t7
python torch_to_pytorch.py --model models/decoder.t7
```

### Test
Use `--content` and `--style` to provide the respective path to the content and style image.
```
CUDA_VISIBLE_DEVICES=<gpu_id> python test.py --content input/content/cornell.jpg --style input/style/woman_with_hat_matisse.jpg
```

You can also run the code on directories of content and style images using `--content_dir` and `--style_dir`. It will save every possible combination of content and styles to the output directory.
```
CUDA_VISIBLE_DEVICES=<gpu_id> python test.py --content_dir input/content --style_dir input/style
```

This is an example of mixing four styles by specifying `--style` and `--style_interpolation_weights` option.
```
CUDA_VISIBLE_DEVICES=<gpu_id> python test.py --content input/content/avril.jpg --style input/style/picasso_self_portrait.jpg,input/style/impronte_d_artista.jpg,input/style/trial.jpg,input/style/antimonocromatismo.jpg --style_interpolation_weights 1,1,1,1 --content_size 512 --style_size 512 --crop
```

Some other options:
* `--content_size`: New (minimum) size for the content image. Keeping the original size if set to 0.
* `--style_size`: New (minimum) size for the content image. Keeping the original size if set to 0.
* `--alpha`: Adjust the degree of stylization. It should be a value between 0.0 and 1.0 (default).
* `--preserve_color`: Preserve the color of the content image.

# python3 test.py --content input/content/lenna.jpg --style input/klmt/eugenia.jpg --decoder models/decoder_klmt_tmp.pth.tar

### Train
Use `--content_dir` and `--style_dir` to provide the respective directory to the content and style images.
```
CUDA_VISIBLE_DEVICES=<gpu_id> python train.py --content_dir <content_dir> --style_dir <style_dir>
```

For more details and parameters, please refer to --help option.

I share the model trained by this code [here](https://drive.google.com/open?id=18wgz9-7sofBocdr2OdRSK2OK19BaQrYA)

## References
- [1]: X. Huang and S. Belongie. "Arbitrary Style Transfer in Real-time with Adaptive Instance Normalization.", in ICCV, 2017.
- [2]: [Original implementation in Torch](https://github.com/xunhuang1995/AdaIN-style)


wiki/HISTORY.md
## decoder 学習時、content と style を入れ替え
あまり意味は無さげ
### saya

wiki/exchange/saya_*.jpg

## style の数を大きく絞る
特異にその style に寄せられる、スタイルが多いと良いとは限らないか
### saya

wiki/mini/saya_*.jpg

## corrector
学習した絵柄かを判断し、復元する。未学習ならば、content の feature を重視、そうでなければ、style の feature を重視

### red
wiki/corrector/red_*.jpg


test.py
import argparse
import os
import torch
import torch.nn as nn
from PIL import Image
from os.path import basename
from os.path import splitext
from torchvision import transforms
from torchvision.utils import save_image

import net
from function import adaptive_instance_normalization, single_adaptive_instance_normalization, correct_adaptive_instance_normalization
from function import coral


def test_transform(size, crop):
    transform_list = []
    if size != 0:
        transform_list.append(transforms.Resize(size))
    if crop:
        transform_list.append(transforms.CenterCrop(size))
    transform_list.append(transforms.ToTensor())
    transform = transforms.Compose(transform_list)
    return transform


def style_transfer(vgg, decoder, abstracter, corrector, content, style, alpha=1.0, interpolation_weights=None):
    assert (0.0 <= alpha <= 1.0)
    content_f = vgg(content)
    #style_gen = vgg(style)
    style_gen = abstracter.execute(content_f) # Attension content -> style
    correct_f = corrector.execute(content_f)
    if interpolation_weights:
        _, C, H, W = content_f.size()
        feat = torch.FloatTensor(1, C, H, W).zero_().to(device)
        base_feat = adaptive_instance_normalization(content_f, style_gen)
        for i, w in enumerate(interpolation_weights):
            feat = feat + w * base_feat[i:i + 1]
        content_f = content_f[0:1]
    else:
        #feat = single_adaptive_instance_normalization(content_f, style_gen)
        feat = correct_adaptive_instance_normalization(content_f, style_gen, correct_f)
    feat = feat * alpha + content_f * (1 - alpha)
    return decoder(feat)


parser = argparse.ArgumentParser()
# Basic options
parser.add_argument('--content', type=str, help='File path to the content image')
parser.add_argument('--content_dir', type=str, help='Directory path to a batch of content images')
parser.add_argument('--style', type=str, help='File path to the style image, or multiple style images separated by commas if you want to do style interpolation or spatial control')
parser.add_argument('--style_dir', type=str, help='Directory path to a batch of style images')
parser.add_argument('--vgg', type=str, default='models/vgg_normalised.pth')
parser.add_argument('--decoder', type=str, required=True)
parser.add_argument('--abstracter', type=str, required=True)
parser.add_argument('--corrector', type=str, required=True)

# Additional options
parser.add_argument('--content_size', type=int, default=512, help='New (minimum) size for the content image, keeping the original size if set to 0')
parser.add_argument('--style_size', type=int, default=512, help='New (minimum) size for the style image, keeping the original size if set to 0')
parser.add_argument('--crop', action='store_true', help='do center crop to create squared image')
parser.add_argument('--save_ext', default='.jpg', help='The extension name of the output image')
parser.add_argument('--output', type=str, default='output', help='Directory to save the output image(s)')

# Advanced options
parser.add_argument('--preserve_color', action='store_true', help='If specified, preserve color of the content image')
parser.add_argument('--alpha', type=float, default=1.0, help='The weight that controls the degree of stylization. Should be between 0 and 1')
parser.add_argument('--style_interpolation_weights', type=str, default='', help='The weight for blending the style of multiple style images')

args = parser.parse_args()

do_interpolation = False

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Either --content or --contentDir should be given.
assert (args.content or args.content_dir)
# Either --style or --styleDir should be given.
#assert (args.style or args.style_dir)

if args.content:
    content_paths = [args.content]
else:
    content_paths = [os.path.join(args.content_dir, f) for f in
                     os.listdir(args.content_dir)]
"""
if args.style:
    style_paths = args.style.split(',')
    if len(style_paths) == 1:
        style_paths = [args.style]
    else:
        do_interpolation = True
        assert (args.style_interpolation_weights != ''), \
            'Please specify interpolation weights'
        weights = [int(i) for i in args.style_interpolation_weights.split(',')]
        interpolation_weights = [w / sum(weights) for w in weights]
else:
    style_paths = [os.path.join(args.style_dir, f) for f in
                   os.listdir(args.style_dir)]
"""
if not os.path.exists(args.output):
    os.mkdir(args.output)

decoder = net.decoder
vgg = net.vgg
abstracter = net.Abstracter2()
corrector = net.Corrector()

decoder.eval()
vgg.eval()
abstracter.eval()
corrector.eval()

decoder.load_state_dict(torch.load(args.decoder))
vgg.load_state_dict(torch.load(args.vgg))
vgg = nn.Sequential(*list(vgg.children())[:31])
abstracter.load(args.abstracter)
corrector.load(args.corrector)

vgg.to(device)
decoder.to(device)
abstracter.to(device)
corrector.to(device)

content_tf = test_transform(args.content_size, args.crop)
style_tf = test_transform(args.style_size, args.crop)

for content_path in content_paths:
    if do_interpolation:  # one content image, N style image
        style = torch.stack([style_tf(Image.open(p)) for p in style_paths])
        content = content_tf(Image.open(content_path)).unsqueeze(0).expand_as(style)
        style = style.to(device)
        content = content.to(device)
        with torch.no_grad():
            output = style_transfer(vgg, decoder, content, style,
                                    args.alpha, interpolation_weights)
        output = output.cpu()
        output_name = '{:s}/{:s}_interpolation{:s}'.format(
            args.output, splitext(basename(content_path))[0], args.save_ext)
        save_image(output, output_name)

    else:  # process one content and one style
        style_paths = ['default']
        for style_path in style_paths:
            content = content_tf(Image.open(content_path))
            #style = style_tf(Image.open(style_path))
            style = None
            #if args.preserve_color: style = coral(style, content)
            #style = style.to(device).unsqueeze(0)
            content = content.to(device).unsqueeze(0)
            with torch.no_grad():
                output = style_transfer(vgg, decoder, abstracter, corrector, content, style, args.alpha)
            output = output.cpu()

            output_name = '{:s}/{:s}_stylized_{:s}{:s}'.format(
                args.output, splitext(basename(content_path))[0],
                splitext(basename(style_path))[0], args.save_ext
            )
            save_image(output, output_name)


net.py
import torch
import torch.nn as nn
import torch.nn.functional as F

from function import adaptive_instance_normalization as adain
from function import calc_mean_std

decoder = nn.Sequential(
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(512, 256, (3, 3)),
    nn.ReLU(),
    nn.Upsample(scale_factor=2, mode='nearest'),
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(256, 256, (3, 3)),
    nn.ReLU(),
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(256, 256, (3, 3)),
    nn.ReLU(),
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(256, 256, (3, 3)),
    nn.ReLU(),
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(256, 128, (3, 3)),
    nn.ReLU(),
    nn.Upsample(scale_factor=2, mode='nearest'),
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(128, 128, (3, 3)),
    nn.ReLU(),
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(128, 64, (3, 3)),
    nn.ReLU(),
    nn.Upsample(scale_factor=2, mode='nearest'),
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(64, 64, (3, 3)),
    nn.ReLU(),
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(64, 3, (3, 3)),
)

vgg = nn.Sequential(
    nn.Conv2d(3, 3, (1, 1)),
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(3, 64, (3, 3)),
    nn.ReLU(),  # relu1-1
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(64, 64, (3, 3)),
    nn.ReLU(),  # relu1-2
    nn.MaxPool2d((2, 2), (2, 2), (0, 0), ceil_mode=True),
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(64, 128, (3, 3)),
    nn.ReLU(),  # relu2-1
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(128, 128, (3, 3)),
    nn.ReLU(),  # relu2-2
    nn.MaxPool2d((2, 2), (2, 2), (0, 0), ceil_mode=True),
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(128, 256, (3, 3)),
    nn.ReLU(),  # relu3-1
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(256, 256, (3, 3)),
    nn.ReLU(),  # relu3-2
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(256, 256, (3, 3)),
    nn.ReLU(),  # relu3-3
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(256, 256, (3, 3)),
    nn.ReLU(),  # relu3-4
    nn.MaxPool2d((2, 2), (2, 2), (0, 0), ceil_mode=True),
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(256, 512, (3, 3)),
    nn.ReLU(),  # relu4-1, this is the last layer used
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(512, 512, (3, 3)),
    nn.ReLU(),  # relu4-2
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(512, 512, (3, 3)),
    nn.ReLU(),  # relu4-3
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(512, 512, (3, 3)),
    nn.ReLU(),  # relu4-4
    nn.MaxPool2d((2, 2), (2, 2), (0, 0), ceil_mode=True),
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(512, 512, (3, 3)),
    nn.ReLU(),  # relu5-1
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(512, 512, (3, 3)),
    nn.ReLU(),  # relu5-2
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(512, 512, (3, 3)),
    nn.ReLU(),  # relu5-3
    nn.ReflectionPad2d((1, 1, 1, 1)),
    nn.Conv2d(512, 512, (3, 3)),
    nn.ReLU()  # relu5-4
)

class Abstracter(nn.Module):
    def __init__(self, size=3):
        super(Abstracter, self).__init__()
        self.size = size
        self.down = nn.Linear(512*(self.size**2), 8)
        self.up = nn.Linear(8, 512*(self.size**2))
        self.relu = nn.ReLU()
        self.dropout = nn.Dropout(0.5)

    def execute(self, input):
        i = nn.functional.upsample(input, size=(self.size,self.size), mode="bilinear", align_corners=True)
        i = i.view(-1, 512*(self.size**2))
        o = self.relu(self.up(self.down(i)))
        mean = o.mean(dim=1)
        std = o.var(dim=1).sqrt() + 0.001
        o = (o - mean) / std + mean
        o = o.view(-1, 512, self.size, self.size)
        return nn.functional.upsample(o, size=input.size()[2:], mode="bilinear", align_corners=True)

    def forward(self, input):
        i = nn.functional.upsample(input, size=(self.size,self.size), mode="bilinear", align_corners=True)
        i = i.view(-1, 512*(self.size**2))
        if self.training:
            i = self.dropout(i)
        o = self.relu(self.up(self.down(i)))
        #print("size=", input.size(), i.size(), o.size())
        return i, o

    def save(self, path):
        state = self.state_dict()
        for key in state.keys():
            state[key] = state[key].to('cpu')
        torch.save(state, path)
    def load(self, path):
        self.load_state_dict(torch.load(path))

class Abstracter2(nn.Module):
    def __init__(self):
        super(Abstracter2, self).__init__()
        self.prepare = nn.Sequential(
            nn.AdaptiveMaxPool2d( (32,32) ))
        self.down = nn.Sequential(
            nn.ReflectionPad2d((1, 1, 1, 1)),
            nn.Conv2d(512, 128, (3, 3)),
            nn.LeakyReLU(),
            nn.MaxPool2d((2, 2), (2, 2), (0, 0), ceil_mode=True),
            nn.ReflectionPad2d((1, 1, 1, 1)),
            nn.Conv2d(128, 32, (3, 3)),
            nn.ReLU(),
            nn.MaxPool2d((2, 2), (2, 2), (0, 0), ceil_mode=True),
            )
        self.up = nn.Sequential(
            #nn.Dropout(0.5),
            nn.Upsample(scale_factor=2, mode='nearest'),
            nn.ReflectionPad2d((1, 1, 1, 1)),
            nn.Conv2d(32, 128, (3, 3)),
            nn.LeakyReLU(),
            nn.Upsample(scale_factor=2, mode='nearest'),
            nn.ReflectionPad2d((1, 1, 1, 1)),
            nn.Conv2d(128, 512, (3, 3)),
            nn.ReLU()
            )

    def execute(self, input):
        size = input.size()[2:]
        i = self.prepare(input)
        m = self.down(i)
        o = self.up(m)
        #print("exe=", i, m, o)
        o = nn.functional.upsample(o, size=size, mode='bilinear', align_corners=True) # あやしい
        #o = nn.functional.upsample(o, size=size, mode='nearest')
        return o.view(1, 512, size[0], size[1])
    def forward(self, input):
        i = self.prepare(input)
        o = self.up(self.down(i))
        return i, o

    def save(self, path):
        state = self.state_dict()
        for key in state.keys():
            state[key] = state[key].to('cpu')
        torch.save(state, path)
    def load(self, path):
        self.load_state_dict(torch.load(path))

class Corrector(nn.Module):
    def __init__(self):
        super(Corrector, self).__init__()
        self.down = nn.Sequential(
            nn.ReflectionPad2d((1, 1, 1, 1)),
            nn.Conv2d(512, 512, (3, 3)),
            nn.ReLU())
        self.correct = nn.Sequential(
            nn.Linear(512, 32),
            nn.Linear(32, 2))
        self.cel = torch.nn.CrossEntropyLoss()
    def execute(self, input):
        size = input.size()
        N, C, H, W = size
        downed = self.down(input)
        corrected = self.correct(downed.transpose(1,3).transpose(1,2).contiguous().view(-1, C))
        cross = F.softmax(corrected * 0.2, dim=1)
        #print(corrected, cross[:,1])
        return cross[:,1].view(N, 1, H, W)
    def forward(self, input):
        size = input.size()
        N, C, H, W = size
        downed = self.down(input)
        tr = downed.transpose(1,3).transpose(1,2).contiguous()
        return self.correct(tr.view(N*H*W, C))

    def save(self, path):
        state = self.state_dict()
        for key in state.keys():
            state[key] = state[key].to('cpu')
        torch.save(state, path)
    def load(self, path):
        self.load_state_dict(torch.load(path))

class Net(nn.Module):
    def __init__(self, encoder, decoder):
        super(Net, self).__init__()
        enc_layers = list(encoder.children())
        self.enc_1 = nn.Sequential(*enc_layers[:4])  # input -> relu1_1
        self.enc_2 = nn.Sequential(*enc_layers[4:11])  # relu1_1 -> relu2_1
        self.enc_3 = nn.Sequential(*enc_layers[11:18])  # relu2_1 -> relu3_1
        self.enc_4 = nn.Sequential(*enc_layers[18:31])  # relu3_1 -> relu4_1
        self.decoder = decoder
        self.mse_loss = nn.MSELoss()

        # fix the encoder
        for name in ['enc_1', 'enc_2', 'enc_3', 'enc_4']:
            for param in getattr(self, name).parameters():
                param.requires_grad = False

    # extract relu1_1, relu2_1, relu3_1, relu4_1 from input image
    def encode_with_intermediate(self, input):
        results = [input]
        for i in range(4):
            func = getattr(self, 'enc_{:d}'.format(i + 1))
            results.append(func(results[-1]))
        return results[1:]

    # extract relu4_1 from input image
    def encode(self, input):
        for i in range(4):
            input = getattr(self, 'enc_{:d}'.format(i + 1))(input)
        return input

    def calc_content_loss(self, input, target):
        assert (input.size() == target.size())
        assert (target.requires_grad is False)
        return self.mse_loss(input, target)

    def calc_style_loss(self, input, target):
        assert (input.size() == target.size())
        assert (target.requires_grad is False)
        input_mean, input_std = calc_mean_std(input)
        target_mean, target_std = calc_mean_std(target)
        return self.mse_loss(input_mean, target_mean) + \
               self.mse_loss(input_std, target_std)

    def forward(self, content, style, alpha=1.0):
        assert 0 <= alpha <= 1
        style_feats = self.encode_with_intermediate(style)
        content_feat = self.encode(content)
        t = adain(content_feat, style_feats[-1])
        t = alpha * t + (1 - alpha) * content_feat

        g_t = self.decoder(t)
        g_t_feats = self.encode_with_intermediate(g_t)

        loss_c = self.calc_content_loss(g_t_feats[-1], t)
        loss_s = self.calc_style_loss(g_t_feats[0], style_feats[0])
        for i in range(1, 4):
            loss_s += self.calc_style_loss(g_t_feats[i], style_feats[i])
        return loss_c, loss_s


function.py
import torch


def calc_mean_std(feat, eps=1e-5):
    # eps is a small value added to the variance to avoid divide-by-zero.
    size = feat.size()
    assert (len(size) == 4)
    N, C = size[:2]
    feat_var = feat.view(N, C, -1).var(dim=2) + eps
    feat_std = feat_var.sqrt().view(N, C, 1, 1)
    feat_mean = feat.view(N, C, -1).mean(dim=2).view(N, C, 1, 1)
    return feat_mean, feat_std

def calc_correct(content_feat, style_feat):
    size = content_feat.size()
    N, C, H, W = size

    content_table = content_feat.transpose(1,3).transpose(1,2).view(-1, C)
    style_table = style_feat.transpose(1,3).transpose(1,2).view(-1, C)
    distance = torch.nn.PairwiseDistance()
    table = distance(content_table, style_table).sqrt()
    table_std = table.var(dim=0).sqrt()
    table_mean = table.mean(dim=0)
    table = torch.clamp(1 - (table - table_mean + table_std * 2) / (table_std * 4), min=0.1, max=1.1).view(N, 1, H, W)
    print("table=", table, table_mean, table_std)

def adaptive_instance_normalization(content_feat, style_feat):
    assert (content_feat.size()[:2] == style_feat.size()[:2])
    size = content_feat.size()
    style_mean, style_std = calc_mean_std(style_feat)
    content_mean, content_std = calc_mean_std(content_feat)

    normalized_feat = (content_feat - content_mean.expand(size)) / content_std.expand(size)
    return normalized_feat * style_std.expand(size) + style_mean.expand(size)

def single_adaptive_instance_normalization(content_feat, style_feat):
    assert (content_feat.size()[:2] == style_feat.size()[:2])
    size = content_feat.size()
    N, C, H, W = size
    style_mean, style_std = calc_mean_std(style_feat)
    normalized_style = (style_feat - style_mean.expand(size)) / style_std.expand(size)
    content_mean, content_std = calc_mean_std(content_feat)
    normalized_content = (content_feat - content_mean.expand(size)) / content_std.expand(size)
    #return style_feat
    #return normalized_content * normalized_style * style_std.expand(size) + style_mean.expand(size)
    #return normalized_content * style_std.expand(size) + style_mean.expand(size)
    return normalized_content * style_std.expand(size) + style_feat

def correct_adaptive_instance_normalization(content_feat, style_feat, correct_feat):
    assert (content_feat.size()[:2] == style_feat.size()[:2])
    size = content_feat.size()
    N, C, H, W = size
    style_mean, style_std = calc_mean_std(style_feat)
    normalized_style = (style_feat - style_mean.expand(size)) / style_std.expand(size)
    content_mean, content_std = calc_mean_std(content_feat)
    normalized_content = (content_feat - content_mean.expand(size)) / content_std.expand(size)
    print("cor=", correct_feat.view(H, W))
    correct = correct_feat.expand(N, C, H, W)
    return (normalized_content + normalized_style * (1.0-correct)) * style_std.expand(size) + style_mean.expand(size)
    #return normalized_content * style_std.expand(size) + style_feat * (1.2 - correct)
    #return normalized_content * (correct + 0.25) * style_std.expand(size) + style_feat # better

def _calc_feat_flatten_mean_std(feat):
    # takes 3D feat (C, H, W), return mean and std of array within channels
    assert (feat.size()[0] == 3)
    assert (isinstance(feat, torch.FloatTensor))
    feat_flatten = feat.view(3, -1)
    mean = feat_flatten.mean(dim=-1, keepdim=True)
    std = feat_flatten.std(dim=-1, keepdim=True)
    return feat_flatten, mean, std


def _mat_sqrt(x):
    U, D, V = torch.svd(x)
    return torch.mm(torch.mm(U, D.pow(0.5).diag()), V.t())


def coral(source, target):
    # assume both source and target are 3D array (C, H, W)
    # Note: flatten -> f

    source_f, source_f_mean, source_f_std = _calc_feat_flatten_mean_std(source)
    source_f_norm = (source_f - source_f_mean.expand_as(
        source_f)) / source_f_std.expand_as(source_f)
    source_f_cov_eye = \
        torch.mm(source_f_norm, source_f_norm.t()) + torch.eye(3)

    target_f, target_f_mean, target_f_std = _calc_feat_flatten_mean_std(target)
    target_f_norm = (target_f - target_f_mean.expand_as(
        target_f)) / target_f_std.expand_as(target_f)
    target_f_cov_eye = \
        torch.mm(target_f_norm, target_f_norm.t()) + torch.eye(3)

    source_f_norm_transfer = torch.mm(
        _mat_sqrt(target_f_cov_eye),
        torch.mm(torch.inverse(_mat_sqrt(source_f_cov_eye)),
                 source_f_norm)
    )

    source_f_transfer = source_f_norm_transfer * \
                        target_f_std.expand_as(source_f_norm) + \
                        target_f_mean.expand_as(source_f_norm)

    return source_f_transfer.view(source.size())


train.py
import os,sys,argparse

sys.path.append('/usr/local/lib/python3.5/dist-packages')
import torch
import torch.backends.cudnn as cudnn
import torch.nn as nn
import torch.utils.data as data
from PIL import Image
from PIL import ImageFile
#from tensorboardX import SummaryWriter
from torchvision import transforms
from tqdm import tqdm
from torch.autograd import Variable

import net
from sampler import InfiniteSamplerWrapper

cudnn.benchmark = True
Image.MAX_IMAGE_PIXELS = None  # Disable DecompressionBombError
ImageFile.LOAD_TRUNCATED_IMAGES = True  # Disable OSError: image file is truncated


def train_transform():
    transform_list = [
        transforms.Resize(size=(512, 512)),
        transforms.RandomCrop(256),
        transforms.ToTensor()
    ]
    return transforms.Compose(transform_list)


class FlatFolderDataset(data.Dataset):
    def __init__(self, root, transform):
        super(FlatFolderDataset, self).__init__()
        self.root = root
        self.paths = os.listdir(self.root)
        self.transform = transform

    def __getitem__(self, index):
        path = self.paths[index]
        img = Image.open(os.path.join(self.root, path)).convert('RGB')
        img = self.transform(img)
        return img

    def __len__(self):
        return len(self.paths)

    def name(self):
        return 'FlatFolderDataset'


def adjust_learning_rate(optimizer, iteration_count):
    """Imitating the original implementation"""
    lr = args.lr / (1.0 + args.lr_decay * iteration_count)
    for param_group in optimizer.param_groups:
        param_group['lr'] = lr


parser = argparse.ArgumentParser()
# Basic options
parser.add_argument('--content_dir', type=str, required=True, help='Directory path to a batch of content images')
parser.add_argument('--style_dir', type=str, required=True, help='Directory path to a batch of style images')
parser.add_argument('--vgg', type=str, default='models/vgg_normalised.pth')
# training options
parser.add_argument('--save_dir', type=str, default='./experiments', help='Directory to save the model')
parser.add_argument('--log_dir', type=str, default='./logs', help='Directory to save the log')
parser.add_argument('--lr', type=float, default=1e-4)
parser.add_argument('--lr_decay', type=float, default=5e-5)
parser.add_argument('--max_iter', type=int, default=160000)
parser.add_argument('--batch_size', type=int, default=10) #  0.6GB / 1batch
parser.add_argument('--style_weight', type=float, default=10.0)
parser.add_argument('--content_weight', type=float, default=1.0)
parser.add_argument('--n_threads', type=int, default=16)
parser.add_argument('--save_model_interval', type=int, default=10000)
parser.add_argument('--decoder', type=str, default=None)
parser.add_argument('--gpu_device', type=int, default=0)
args = parser.parse_args()

if torch.cuda.is_available():
    device = 'cuda'
    torch.cuda.set_device(args.gpu_device)
else:
    device = 'cpu'
    print("ATTENSION: CPU mode.")

if not os.path.exists(args.save_dir):
    os.mkdir(args.save_dir)

if not os.path.exists(args.log_dir):
    os.mkdir(args.log_dir)
#writer = SummaryWriter(log_dir=args.log_dir)

decoder = net.decoder
vgg = net.vgg
abstracter = net.Abstracter2()
abstracter.train()
abstracter.to(device)
corrector = net.Corrector()
corrector.train()
corrector.to(device)

is_decoder_updating = True
is_abstracter_updating = True

if args.decoder is not None and os.path.exists(args.decoder):
    decoder.load_state_dict(torch.load(args.decoder))
    is_decoder_updating = False
    print("decoder is loaded.")
vgg.load_state_dict(torch.load(args.vgg))
vgg = nn.Sequential(*list(vgg.children())[:31])
network = net.Net(vgg, decoder)
network.train()
network.to(device)

content_tf = train_transform()
style_tf = train_transform()

content_dataset = FlatFolderDataset(args.content_dir, content_tf)
style_dataset = FlatFolderDataset(args.style_dir, style_tf)

content_iter = iter(data.DataLoader(
    content_dataset, batch_size=args.batch_size,
    sampler=InfiniteSamplerWrapper(content_dataset),
    num_workers=args.n_threads))
style_iter = iter(data.DataLoader(
    style_dataset, batch_size=args.batch_size,
    sampler=InfiniteSamplerWrapper(style_dataset),
    num_workers=args.n_threads))

optimizer = torch.optim.Adam(network.decoder.parameters(), lr=args.lr)
optimizer_a = torch.optim.Adam(abstracter.parameters())
criterion_a = nn.MSELoss()
optimizer_c = torch.optim.Adam(corrector.parameters())
criterion_c = nn.CrossEntropyLoss()

for i in tqdm(range(args.max_iter)):
    adjust_learning_rate(optimizer, iteration_count=i)
    content_images = next(content_iter).to(device)
    style_images = next(style_iter).to(device)

    # decoder
    if is_decoder_updating:
        loss_c, loss_s = network(content_images, style_images)
        loss_c = args.content_weight * loss_c
        loss_s = args.style_weight * loss_s
        loss = loss_c + loss_s
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

    # abstracter
    if is_abstracter_updating:
        content_enc = network.encode(content_images)
        style_enc = network.encode(style_images)
        _i, _o = abstracter(style_enc)
        loss_a = criterion_a(_o, _i)
        optimizer_a.zero_grad()
        loss_a.backward()
        optimizer_a.step()

        N, C, H, W = style_enc.size()
        ones = torch.ones([N*H*W], dtype=torch.long, device=device)
        zeros = torch.zeros([N*H*W], dtype=torch.long, device=device)
        loss_c_1 = criterion_c(corrector(style_enc), ones)        
        loss_c_0 = criterion_c(corrector(content_enc), zeros)
        loss_c = loss_c_0 + loss_c_1
        optimizer_c.zero_grad()
        loss_c.backward()
        optimizer_c.step()

    #writer.add_scalar('loss_content', loss_c.item(), i + 1)
    #writer.add_scalar('loss_style', loss_s.item(), i + 1)

    # saving
    if (i + 1) % args.save_model_interval == 0 or (i + 1) == args.max_iter:
        if is_decoder_updating:
            state_dict = net.decoder.state_dict()
            for key in state_dict.keys():
                state_dict[key] = state_dict[key].to(torch.device('cpu'))
            torch.save(state_dict, '{:s}/decoder_iter_{:d}.pth.tar'.format(args.save_dir, i + 1))
        if is_abstracter_updating:
            abstracter.save('{:s}/abstracter_{:d}.pth.tar'.format(args.save_dir, i + 1))
            corrector.save('{:s}/corrector_{:d}.pth.tar'.format(args.save_dir, i + 1))
        
#writer.close()
